sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java

```diff
@@ -529,4 +534,49 @@ public String toString() {
       return Collections.<TupleTag<?>, PValue>singletonMap(tag, pCollection);
     }
   }
+
+  /** Accessor for the value of a side input in the current context. */
+  @Internal
+  public interface SideInputContext {
+    <T> T sideInput(PCollectionView<T> view);
+  }
+
+  private static ThreadLocal<SideInputContext> sideInputContextThreadLocal = new ThreadLocal<>();
+
+  /**
+   * Result of {@link #setCurrentSideInputContext(SideInputContext)}. Should be used only by
+   * runners.
+   */
+  @Internal
+  public interface ScopedSideInputContext extends AutoCloseable {
+    @Override
+    void close();
+  }
+
+  /**
+   * Sets the thread-local {@link SideInputContext} and restores the old one when closed. Should be
+   * used only by runners.
+   */
+  @Internal
+  public static ScopedSideInputContext setCurrentSideInputContext(SideInputContext context) {
+    final SideInputContext old = sideInputContextThreadLocal.get();
+    sideInputContextThreadLocal.set(context);
+    return new ScopedSideInputContext() {
+      @Override
+      public void close() {
+        sideInputContextThreadLocal.set(old);
+      }
+    };
+  }
+
+  /** Returns the current {@link SideInputContext}. Should be used only by runners. */
+  @Internal
+  public static SideInputContext getCurrentSideInputContext() {
+    SideInputContext context = sideInputContextThreadLocal.get();
+    if (context == null) {
+      throw new IllegalStateException(
+          "Current thread is not in a context where side inputs are available");
```

Comment:
<!--new_thread; commit:2b7b6939227e3e105feee6eddf1d6c943c40815a; resolved:0-->
Should we specifically mention processElement?